src/ui_basic/dropdown.h

```diff
@@ -275,9 +287,78 @@ template <typename Entry> class Dropdown : public BaseDropdown {
 	                  type,
 	                  style,
 	                  button_style) {
+		set_handle_textinput();
 	}
 	~Dropdown() override {
-		entry_cache_.clear();
+		filtered_entries.clear();
+		unfiltered_entries.clear();
+	}
+
+	void clear_filter() override {
+		if (current_filter_.empty()) {
+			return;
+		}
+		current_filter_.clear();
+		if (has_selection()) {
+			Entry selected_entry = get_selected();
+			restore_filtered_list();
+			select(selected_entry);
+		} else {
+			restore_filtered_list();
+		}
+	}
+	void restore_filtered_list() {
+		clear_filtered_list();
+		for (auto& ee : unfiltered_entries) {
+			add_to_filtered_list(ee.name, ee.value, ee.img, false, ee.tooltip, ee.hotkey);
+		}
+	}
+
+	void save_selected_entry(uint32_t index) override {
+		selected_entry_ = index < filtered_entries.size() ? *filtered_entries[index] : Entry{};
+	}
+
+	bool handle_textinput(const std::string& input_text) override {
+		std::string lower = std::string(input_text);
+		lower[0] = std::tolower(lower[0]);
+		current_filter_ = current_filter_.append(lower);
+
+		apply_filter();
+
+		return true;
+	}
+
+	void apply_filter() override {
+		clear_filtered_list();
+		add_matching_entries();
+
+		if (filtered_entries.empty()) {
+			add_no_match_entry();
+		}
+		// force list to stay open even if mouse is now away due to smaller
+		// dropdown list because of applied filter
+		set_list_visibility(true);
+	}
+
+	void add_no_match_entry() {
+		// re-add initially selected entry with adapted texts to inform user
+		for (auto& x : unfiltered_entries) {
+			if (x.value == selected_entry_) {
+				add_to_filtered_list(_("No matches"), x.value, x.img, false,
+				                     _("Previously selected entry will be selected again"), x.hotkey);
```

Comment:
IMHO this message just confuses the user as it should be obvious that clicking the "No matches" entry doesn't change anything. Better leave the tooltip empty.

Reply:
Actually I am more in favor of deleting depth as it is only written, never read. Maybe it was used once upon a time but is not anymore? What is your opinion?

Reply:
hmm, I found this very smart :D I thought I might not be sure what if I can expect on an element which says "no matches". Does this tooltip hurt so much?
If you insist, I will delete it...

Reply:
How about an empty list item with a tooltip that says "No matches"?

Reply:
might be a good idea, pictorial dropdowns look awkward though. Setting image to nullptr (resulting in a non-displayed entry also for pictorial dropdowns) might cause subtle bugs later on (we don't allow entries with no picture for pictorial dropdowns - see line 300 in dropdown.cc)
![Selection_003](https://user-images.githubusercontent.com/10723431/131264586-910e06d4-d762-4f4b-858d-29784a341ccd.png)

Reply:
You can use `data/images/wui/editor/no_ware.png`